File: Unidad 1/Imagenes con OpenCV/Histogram.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <opencv2/opencv.hpp>
using namespace std;
using namespace cv;

vector<int> generate_histogram(const Mat& image) {
    vector<int> histogram(256, 0);
    int rows = image.rows;
    int columns = image.cols;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            int pixel_value = image.at<uchar>(i, j);
            histogram[pixel_value]++;
        }
    }
    return histogram;;
}

void save_histogram(const vector<int>& histogram, const string& filename) {
    ofstream file(filename);
    for (size_t i = 0; i < histogram.size(); i++) {
        file << i << " " << histogram[i] << endl;
    }
    file.close();
}

Mat equalization(const Mat& image, const vector<int>& histogram) {
    Mat image_equalized = image.clone();
    int total_pixels = image.rows * image.cols;
    vector<uchar> f(256);
    
    int acumulado = histogram[0];
    f[0] = 0;
    for (int i = 1; i < 255; i++) {
        f[i] = (uchar)((255.0f * acumulado) / total_pixels);
        acumulado += histogram[i];
    }
    f[255] = 255;
    
    for (int i = 0; i < image.rows; i++) {
        for (int j = 0; j < image.cols; j++) {
            uchar pixel_value = image.at<uchar>(i, j);
            image_equalized.at<uchar>(i, j) = f[pixel_value];
        }
    }

    return image_equalized;
}

int main() {
    string image;
    cout << "Enter the name file image (with extension): ";
    cin >> image;
    string path_image = "D:/UNSA EPCC/7mo semestre/Computacion Grafica/Unidad 1/Imagenes con OpenCV/Imagenes/Equalization Histogram Images/" + image;
    string path_histogram = "D:/UNSA EPCC/7mo semestre/Computacion Grafica/Unidad 1/Imagenes con OpenCV/Imagenes/Equalization Histogram Images/";
    Mat image_original = imread(path_image, IMREAD_GRAYSCALE);
    
    if (image_original.empty()) {
        cerr << "Error loading image: " << path_image << endl;
        return -1;
    }

    imshow("Image Original", image_original);

    vector<int> histogram = generate_histogram(image_original);
    save_histogram(histogram, path_histogram + "histogram.txt");

    Mat image_equalized = equalization(image_original, histogram);
    imshow("Image Equalized", image_equalized);

    waitKey(0);
    return 0;
}

File: Unidad 1/Imagenes con OpenCV/Histogram.py
import matplotlib.pyplot as plt
import numpy as np
import os

def display_histogram(filename, name_histogram, output_folder):
    dates = np.loadtxt(filename)
    intensitys = dates[:, 0]
    frecuencys = dates[:, 1]
    
    plt.figure(figsize=(10, 5))
    plt.fill_between(intensitys, frecuencys, color='black', alpha=0.8)
    
    plt.yscale('log')
    plt.title(name_histogram)
    plt.xlabel('Intensity')
    plt.ylabel('Frequency (log scale)')
    plt.grid(True, which='both', linestyle='--', linewidth=0.5)
    
    plt.tight_layout()
    output_path = os.path.join(output_folder, name_histogram + '_filled.png')
    plt.savefig(output_path)
    plt.show()

name_file = input("Enter the name of the file: ")
input_folder = "D:/UNSA EPCC/7mo semestre/Computacion Grafica/Unidad 1/Imagenes con OpenCV/Imagenes/Equalization Histogram Images/"
output_folder = "D:/UNSA EPCC/7mo semestre/Computacion Grafica/Unidad 1/Imagenes con OpenCV/Imagenes/Equalization Histogram Images/Histograms/"

file_original = os.path.join(input_folder, f"histogram_{name_file}_original.txt")
file_equalized = os.path.join(input_folder, f"histogram_{name_file}_equalized.txt")
file_binarized = os.path.join(input_folder, f"histogram_{name_file}_binarized.txt")

display_histogram(file_original, f"Histogram {name_file} Original", output_folder)
display_histogram(file_equalized, f"Histogram {name_file} Equalized", output_folder)
display_histogram(file_binarized, f"Histogram {name_file} Binarized", output_folder)